File: velolabs_repos/lattis_android/app/src/main/java/com/lattis/ellipse/data/database/base/AbstractRealmDataMapper.java
package com.lattis.ellipse.data.database.base;

import androidx.annotation.NonNull;

import java.util.ArrayList;
import java.util.List;

import io.realm.RealmModel;
import io.realm.RealmResults;

public abstract class AbstractRealmDataMapper<IN,OUT extends RealmModel> {


    @NonNull
    public List<OUT> mapIn(@NonNull List<IN> objects) {
        List<OUT> realmObjects = new ArrayList<>();
        for (IN object : objects) {
            realmObjects.add(mapIn(object));
        }
        return realmObjects;
    }

    @NonNull
    public abstract OUT mapIn(@NonNull IN object);

    @NonNull
    public List<IN> mapOut(@NonNull List<OUT> realmObjectList) {
        List<IN> objects = new ArrayList<>();
        for (OUT realObject : realmObjectList) {
            objects.add(mapOut(realObject));
        }
        return objects;
    }

    @NonNull
    public List<IN> mapOut(@NonNull RealmResults<OUT> realmObjectList) {
        List<IN> objects = new ArrayList<>();
        for (OUT realObject : realmObjectList) {
            objects.add(mapOut(realObject));
        }
        return objects;
    }

    @NonNull
    public abstract IN mapOut(@NonNull OUT realmObject);
}

File: velolabs_repos/lattis_android_v2/data/src/main/java/com/lattis/data/database/base/AbstractRealmDataMapper.kt
package com.lattis.data.database.base

import java.util.ArrayList

import io.realm.RealmModel
import io.realm.RealmResults

abstract class AbstractRealmDataMapper<IN, OUT : RealmModel> {


    fun mapIn(objects: List<IN>): List<OUT> {
        val realmObjects = ArrayList<OUT>()
        for (`object` in objects) {
            realmObjects.add(mapIn(`object`))
        }
        return realmObjects
    }

    abstract fun mapIn(`object`: IN): OUT

    fun mapOut(realmObjectList: List<OUT>): List<IN> {
        val objects = ArrayList<IN>()
        for (realObject in realmObjectList) {
            objects.add(mapOut(realObject))
        }
        return objects
    }

    open fun mapOut(realmObjectList: RealmResults<OUT>): List<IN> {
        val objects = ArrayList<IN>()
        for (realObject in realmObjectList) {
            objects.add(mapOut(realObject))
        }
        return objects
    }

    abstract fun mapOut(realmObject: OUT): IN
}
